File: bindings/java/src/main/java/io/omq/RecvRing.java
package io.omq;

import java.lang.foreign.MemorySegment;
import java.lang.foreign.ValueLayout;
import java.lang.invoke.VarHandle;
import java.nio.BufferOverflowException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.ReadOnlyBufferException;
import java.util.Optional;

final class RecvRing implements AutoCloseable {
    private static final int DEFAULT_DESC_CAPACITY = 1024;
    private static final long DEFAULT_PAYLOAD_CAPACITY = 4L * 1024L * 1024L;
    private static final int FILL_BATCH = 256;

    private static final long CONTROL_BYTES = 256;
    private static final long CONTROL_HEAD = 0;
    private static final long CONTROL_TAIL = 128;

    private static final long DESC_BYTES = 64;
    private static final long DESC_PAYLOAD = 0;
    private static final long DESC_PAYLOAD_LEN = 8;
    private static final long DESC_PART_COUNT = 24;
    private static final long DESC_FLAGS = 32;

    private static final long FLAG_MULTIPART = 1;
    private static final long FLAG_EXTERNAL = 2;

    private static final ValueLayout.OfInt INT =
            ValueLayout.JAVA_INT_UNALIGNED.withOrder(ByteOrder.nativeOrder());
    private static final ValueLayout.OfLong LONG =
            ValueLayout.JAVA_LONG.withOrder(ByteOrder.nativeOrder());
    private static final VarHandle ATOMIC_LONG = LONG.varHandle();

    private long handle;
    private MemorySegment control;
    private MemorySegment descriptors;
    private MemorySegment payload;
    private long descMask;
    private long head;
    private long releasedHead;
    private long cachedTail;

    Message receive(long socketHandle, long timeoutMillis) {
        ensure(socketHandle);
        Desc desc = next(timeoutMillis);
        try {
            return readMessage(desc);
        } finally {
            advance();
        }
    }

    byte[] receiveBytes(long socketHandle, long timeoutMillis) {
        ensure(socketHandle);
        Desc desc = next(timeoutMillis);
        try {
            if (desc.partCount() != 1) {
                throw new IllegalStateException("message has " + desc.partCount() + " parts");
            }
            return readBytes(desc);
        } finally {
            advance();
        }
    }

    int receiveInto(long socketHandle, ByteBuffer destination, long timeoutMillis) {
        ensure(socketHandle);
        Desc desc = next(timeoutMillis);
        try {
            if (desc.partCount() != 1) {
                throw new IllegalStateException("message has " + desc.partCount() + " parts");
            }
            if (destination.isReadOnly()) {
                throw new ReadOnlyBufferException();
            }
            int len = checkedIntLength(desc.payloadLen());
            if (len > destination.remaining()) {
                throw new BufferOverflowException();
            }
            if (len > 0) {
                MemorySegment.copy(
                        source(desc), sourceOffset(desc),
                        MemorySegment.ofBuffer(destination), 0,
                        len);
            }
            destination.position(destination.position() + len);
            return len;
        } finally {
            advance();
        }
    }

    Optional<Message> tryReceiveCachedMessage() {
        if (handle == 0 || !hasCached()) {
            return Optional.empty();
        }
        Desc desc = current();
        try {
            return Optional.of(readMessage(desc));
        } finally {
            advance();
        }
    }

    private Desc next(long timeoutMillis) {
        if (!hasCached()) {
            releaseConsumed();
            NativeFfm.recvRingFill(handle, timeoutMillis, FILL_BATCH);
            cachedTail = tailAcquire();
        }
        if (!hasCached()) {
            throw new TimeoutException("operation timed out");
        }
        return current();
    }

    private boolean hasCached() {
        if (head != cachedTail) {
            return true;
        }
        cachedTail = tailAcquire();
        return head != cachedTail;
    }

    private Desc current() {
        long offset = (head & descMask) * DESC_BYTES;
        return new Desc(
                descriptors.get(LONG, offset + DESC_PAYLOAD),
                descriptors.get(LONG, offset + DESC_PAYLOAD_LEN),
                descriptors.get(LONG, offset + DESC_PART_COUNT),
                descriptors.get(LONG, offset + DESC_FLAGS));
    }

    private Message readMessage(Desc desc) {
        if (desc.partCount() == 1) {
            return Message.fromNative(readBytes(desc));
        }
        return Message.fromNative(readParts(desc));
    }

    private byte[] readBytes(Desc desc) {
        int len = checkedIntLength(desc.payloadLen());
        byte[] out = new byte[len];
        if (len > 0) {
            MemorySegment.copy(source(desc), sourceOffset(desc), MemorySegment.ofArray(out), 0, len);
        }
        return out;
    }

    private byte[][] readParts(Desc desc) {
        if ((desc.flags() & FLAG_MULTIPART) == 0) {
            return new byte[][] {readBytes(desc)};
        }
        MemorySegment source = source(desc);
        long offset = sourceOffset(desc);
        int partCount = checkedIntLength(desc.partCount());
        int encodedPartCount = source.get(INT, offset);
        if (encodedPartCount != partCount) {
            throw new OMQException("native receive ring metadata mismatch");
        }
        byte[][] parts = new byte[partCount][];
        long bodyOffset = offset + 4L + 4L * partCount;
        for (int i = 0; i < partCount; i++) {
            int len = source.get(INT, offset + 4L + 4L * i);
            if (len < 0) {
                throw new OMQException("native receive ring part length overflow");
            }
            byte[] part = new byte[len];
            if (len > 0) {
                MemorySegment.copy(source, bodyOffset, MemorySegment.ofArray(part), 0, len);
            }
            parts[i] = part;
            bodyOffset += len;
        }
        return parts;
    }

    private MemorySegment source(Desc desc) {
        if ((desc.flags() & FLAG_EXTERNAL) != 0) {
            return MemorySegment.ofAddress(desc.payload()).reinterpret(desc.payloadLen());
        }
        return payload;
    }

    private long sourceOffset(Desc desc) {
        return (desc.flags() & FLAG_EXTERNAL) != 0 ? 0 : desc.payload();
    }

    private void advance() {
        head++;
        if (head == cachedTail) {
            releaseConsumed();
        }
    }

    private long tailAcquire() {
        return (long) ATOMIC_LONG.getAcquire(control, CONTROL_TAIL);
    }

    private void releaseConsumed() {
        if (releasedHead != head) {
            ATOMIC_LONG.setRelease(control, CONTROL_HEAD, head);
            releasedHead = head;
        }
    }

    private void ensure(long socketHandle) {
        if (handle != 0) {
            return;
        }
        long created = NativeFfm.recvRingCreate(
                socketHandle, DEFAULT_DESC_CAPACITY, DEFAULT_PAYLOAD_CAPACITY);
        int descCapacity = NativeFfm.recvRingDescCapacity(created);
        long payloadCapacity = NativeFfm.recvRingPayloadCapacity(created);
        long controlAddress = NativeFfm.recvRingControlAddress(created);
        long descAddress = NativeFfm.recvRingDescAddress(created);
        long payloadAddress = NativeFfm.recvRingPayloadAddress(created);
        if (descCapacity <= 0
                || !isPowerOfTwo(descCapacity)
                || payloadCapacity <= 0
                || !isPowerOfTwo(payloadCapacity)
                || controlAddress == 0 || descAddress == 0 || payloadAddress == 0) {
            NativeFfm.recvRingClose(created);
            throw new OMQException("native receive ring returned invalid memory");
        }
        handle = created;
        control = MemorySegment.ofAddress(controlAddress).reinterpret(CONTROL_BYTES);
        descriptors = MemorySegment.ofAddress(descAddress)
                .reinterpret((long) descCapacity * DESC_BYTES);
        payload = MemorySegment.ofAddress(payloadAddress).reinterpret(payloadCapacity);
        descMask = descCapacity - 1L;
        head = 0;
        releasedHead = 0;
        cachedTail = 0;
    }

    @Override
    public void close() {
        long current = handle;
        if (current == 0) {
            return;
        }
        handle = 0;
        control = null;
        descriptors = null;
        payload = null;
        NativeFfm.recvRingClose(current);
    }

    private static int checkedIntLength(long len) {
        if (len < 0 || len > Integer.MAX_VALUE) {
            throw new OMQException("message is too large for a Java byte array");
        }
        return (int) len;
    }

    private static boolean isPowerOfTwo(long value) {
        return value > 0 && (value & (value - 1)) == 0;
    }

    private record Desc(
            long payload,
            long payloadLen,
            long partCount,
            long flags) {
    }
}

File: bindings/go/recv_ring.go
package omq

import (
	"sync/atomic"
	"unsafe"
)

const (
	defaultRecvRingDescCapacity    = 1024
	defaultRecvRingPayloadCapacity = 4 * 1024 * 1024
	recvRingFillBatch              = 512

	recvRingControlHead = 0
	recvRingControlTail = 128

	recvRingFlagMultipart = 1
	recvRingFlagExternal  = 2
)

type recvRingDesc struct {
	payload    uint64
	payloadLen uint64
	totalLen   uint64
	partCount  uint64
	flags      uint64
	payloadEnd uint64
	_          [2]uint64
}

type recvRing struct {
	handle       *nativeRecvRing
	control      unsafe.Pointer
	descriptors  []recvRingDesc
	payload      []byte
	descMask     uint64
	head         uint64
	releasedHead uint64
	cachedTail   uint64
	viewActive   bool
}

func newRecvRing(socket *nativeSocket, ringSize int) (*recvRing, error) {
	descCapacity := defaultRecvRingDescCapacity
	if ringSize > 0 {
		descCapacity = ringSize
	}
	handle, memory, err := recvRingCreateNative(
		socket,
		descCapacity,
		defaultRecvRingPayloadCapacity,
	)
	if err != nil {
		return nil, err
	}
	if memory.control == nil || memory.descriptors == nil || memory.payload == nil ||
		memory.descCapacity <= 0 || memory.payloadCapacity <= 0 ||
		!isPowerOfTwo(uint64(memory.descCapacity)) ||
		!isPowerOfTwo(uint64(memory.payloadCapacity)) {
		recvRingCloseNative(handle)
		return nil, &ConfigError{Err: "native receive ring returned invalid memory"}
	}
	return &recvRing{
		handle:      handle,
		control:     memory.control,
		descriptors: unsafe.Slice((*recvRingDesc)(memory.descriptors), memory.descCapacity),
		payload:     unsafe.Slice((*byte)(memory.payload), memory.payloadCapacity),
		descMask:    uint64(memory.descCapacity - 1),
	}, nil
}

func (r *recvRing) tryRecvView() (RecvView, error) {
	if r == nil || r.handle == nil {
		return RecvView{}, ErrClosed
	}
	r.releaseActiveView()
	if err := r.fillIfEmpty(0); err != nil {
		return RecvView{}, err
	}
	desc := r.current()
	if desc.partCount != 1 {
		r.advance()
		return RecvView{}, &ConfigError{Err: "RecvView requires a single-part message"}
	}
	if desc.flags&recvRingFlagExternal != 0 {
		r.advance()
		return RecvView{}, &ConfigError{Err: "RecvView payload exceeds receive ring capacity"}
	}
	r.viewActive = true
	return RecvView{data: r.source(desc)}, nil
}

func (r *recvRing) tryRecvInto(dst []byte) (int, error) {
	return r.recvInto(dst, 0)
}

func (r *recvRing) recvIntoBlocking(dst []byte) (int, error) {
	return r.recvInto(dst, -1)
}

func (r *recvRing) recvInto(dst []byte, timeoutMillis int64) (int, error) {
	if r == nil || r.handle == nil {
		return 0, ErrClosed
	}
	r.releaseActiveView()
	if err := r.fillIfEmpty(timeoutMillis); err != nil {
		return 0, err
	}
	desc := r.current()
	defer r.advance()
	if desc.partCount != 1 {
		return 0, &ConfigError{Err: "RecvInto requires a single-part message"}
	}
	if desc.flags&recvRingFlagExternal != 0 {
		return 0, &ConfigError{Err: "RecvInto payload exceeds receive ring capacity"}
	}
	if desc.payloadLen > uint64(len(dst)) {
		return 0, &MessageTooLargeError{Err: "destination buffer too small"}
	}
	body := r.source(desc)
	copy(dst, body)
	return len(body), nil
}

func (r *recvRing) close() {
	if r == nil || r.handle == nil {
		return
	}
	r.releaseActiveView()
	r.releaseConsumed()
	recvRingCloseNative(r.handle)
	r.handle = nil
	r.control = nil
	r.descriptors = nil
	r.payload = nil
}

func (r *recvRing) fillIfEmpty(timeoutMillis int64) error {
	if r.hasCached() {
		return nil
	}
	r.releaseConsumed()
	if err := recvRingFillNative(r.handle, timeoutMillis, recvRingFillBatch); err != nil {
		return err
	}
	r.cachedTail = r.tailAcquire()
	if !r.hasCached() {
		return ErrAgain
	}
	return nil
}

func (r *recvRing) hasCached() bool {
	if r.head != r.cachedTail {
		return true
	}
	r.cachedTail = r.tailAcquire()
	return r.head != r.cachedTail
}

func (r *recvRing) current() recvRingDesc {
	return r.descriptors[r.head&r.descMask]
}

func (r *recvRing) source(desc recvRingDesc) []byte {
	if desc.payloadLen == 0 {
		return nil
	}
	return r.payload[desc.payload : desc.payload+desc.payloadLen]
}

func (r *recvRing) releaseActiveView() {
	if r.viewActive {
		r.advance()
		r.viewActive = false
	}
}

func (r *recvRing) advance() {
	r.head++
	if r.head == r.cachedTail {
		r.releaseConsumed()
	}
}

func (r *recvRing) releaseConsumed() {
	if r.releasedHead == r.head {
		return
	}
	atomic.StoreUint64(r.headPtr(), r.head)
	r.releasedHead = r.head
}

func (r *recvRing) tailAcquire() uint64 {
	return atomic.LoadUint64(r.tailPtr())
}

func (r *recvRing) headPtr() *uint64 {
	return (*uint64)(unsafe.Add(r.control, recvRingControlHead))
}

func (r *recvRing) tailPtr() *uint64 {
	return (*uint64)(unsafe.Add(r.control, recvRingControlTail))
}
